### PR TITLE
Update property enums to support CodepointTrie

### LINKED ISSUE #1071: Define property enums/newtypes, including AsULE impls

### PR DESCRIPTION
**Pull request by iainireland**
Putting up an initial draft for feedback. This partially resolves #1071: it handles General_Category, but not Script yet.

Points of interest:
1. Going along with a discussion we had earlier, I split GeneralCategory into GeneralCategory and GeneralSubcategory. The former is `[repr(u32)]` and can represent grouped categories like `Letter`. It's useful for queries. The latter is `[repr(u8)]` and represents a single specific subcategory (eg TitlecaseLetter). It's primarily useful for output. GeneralSubcategory is what's stored in the codepoint trie.
2. I pulled in `num_enum` as a dependency to automatically derive `GeneralSubcategory::try_from::<u8>`. It is no_std compatible, has 2.5M downloads all time, and has minimal additional dependencies. The alternative would be writing a couple of large match statements by hand, or writing our own proc_macro to do so.
3. In retrospect it's possible that icu_uniset should instead have been icu_uprops or something like that, since the same property definitions are used for both Uniset and CodepointTrie. For now I've put the ULE code in a separate file in icu_uniset; we can sort things out more later.

@echeran @sffc @Manishearth

### CODE REVIEW

**Review by Manishearth — APPROVED** [review]
ULE code looks correct,

**Review by sffc — COMMENTED** [review]
The GeneralSubcategoryULE looks about right! Although you need to take in the latest changes to the trait from recent PRs.

**Comment by iainireland** [comment]
Assuming that CI doesn't have any surprises for me, this version is ready for review. I think this resolves #1071; if there's anything that was supposed to be included that isn't here, let me know.

I've gone ahead and converted Script to a newtype-wrapped u16; it's a bit ugly, but manageably so, and the difference between an enum pretending to be a u16 and a u16 pretending to be an enum is small.

File: components/uniset/src/enum_props.rs
Comment (echeran) [inline components/uniset/src/enum_props.rs:27]: Keep in mind the tradeoffs being made by this `GeneralSubcategory` + `GeneralCategory` duo of enums. As I currently see them:

Pros:
* you have an enum that also behaves like a bitset with no extra deps
* this is probably fine in the return-a-UnicodeSet usage pattern (ex: regex)

Cons:
* for the get-property-value-at-this-code-point usage pattern, you still won't return any of the multi-category variants as the return value even if it's a `GeneralCategory` enum, so the code to test for "was that value in the Letter multi-category?" still may need to be a predicate function, but now you have a caveat that the set of valid values for `GeneralCategory` varies depending on the context.

Comment (iainireland) [reply components/uniset/src/enum_props.rs:27]: I'm not sure what you mean by "predicate function" here. We can write
```rust
fn is_char_in_general_category(c: char, gc: GeneralCategory) -> bool {
  let sc: GeneralSubcategory = get_subcategory(c);
  1 << (sc as u32) & gc as u32 != 0
}
```
In general, I think the rule of thumb is that we use GeneralCategory for inputs, and GeneralSubcategory for outputs.

File: components/uniset/src/enum_props.rs
Comment (sffc) [inline components/uniset/src/enum_props.rs:28]: Nit 1: I think you should keep the `#[non_exhaustive]`

Nit 2 (optional): I would argue that this isn't really an enum; it's a set of **selectors** for common cases.  If it were me, these would be constants wrapped in a newtype.  But I don't feel strongly enough on that point to hold up the PR.

File: components/uniset/src/enum_props.rs
Comment (sffc) [inline components/uniset/src/enum_props.rs:159]: Nit (optional): I still don't see the use case behind defining all these constants.

File: components/uniset/src/props.rs
Comment (sffc) [inline components/uniset/src/props.rs:718]: Thought: This just means that you can't proactively request a particular UnicodeSet for a script that wasn't known at build time.  That is smelly, but I'll accept it.  If you really want to access the data for a new script from an old version of ICU4X, you should be able to get it from the CodePointTrie API.

File: components/uniset/src/ule.rs
Comment (sffc) [inline components/uniset/src/ule.rs:53]: Nit: remove these two methods since they are now deafult-implemented

File: components/uniset/src/ule.rs
Comment (sffc) [inline components/uniset/src/ule.rs:57]: Thought: I had assumed we would have a `ScriptULE` here, but you are correct that it would be the same as `PlainOldULE<2>`, so we should use the existing type instead.

**Review by sffc — CHANGES_REQUESTED** [review]
Mostly looks great!  Just a few minor things

File: components/uniset/src/enum_props.rs
Comment (sffc) [reply components/uniset/src/enum_props.rs:27]: The "predicate function" could be

```rust
impl GeneralCategory {
    pub fn contains(sub: GeneralSubcategory) { ... }
}
```

File: components/uniset/src/props.rs
Comment (iainireland) [reply components/uniset/src/props.rs:718]: 1. Proactive requests for currently unknown scripts are the long tail of the long tail, I think. There's not a lot of demand for Vithquki.
2. This restriction is driven by the way we represent our resource keys: you can't proactively request an unknown script because we don't have any way to retrieve it. If we refactored that, we could also rewrite this function to provide the data. The API doesn't have to change; Script can already represent unknown scripts. If there's demand, we can support it.

File: components/uniset/src/enum_props.rs
Comment (iainireland) [reply components/uniset/src/enum_props.rs:159]: I use these constants in `icu4x_js_regexp` (and also in `get_script_val_set`). I think a) there's a clear case for defining the more common constants, b) there's no clear line for how we would decide which scripts aren't important enough to define, and c) the cost of defining a constant is roughly zero, so we should err on the side of generosity to the user.

File: components/uniset/src/enum_props.rs
Comment (iainireland) [reply components/uniset/src/enum_props.rs:28]: It's true that in principle we could have a different implementation that allows users to select custom sets of subcategories (say, Letter+Number). However: 
1. Our current data model - that is, the way we define our resource keys - can't provide data for a selector we haven't defined. 
2. I think that usecase is easily satisfied by just making multiple queries.
3. If we keep this marked as non-exhaustive, we can switch later.

Comment (iainireland) [reply components/uniset/src/enum_props.rs:28]: I went to write `impl From<GeneralSubcategory> for GeneralCategory`, and now I lean towards making this a newtype-wrapped selector, because it means the `from` impl is just a bitshift instead of a match.

File: components/uniset/src/ule.rs
Comment (Manishearth) [inline components/uniset/src/ule.rs:30]: issue: Please add a comment explaining why this impl satisfied the invariants of ULE

(yes, I know that the ZeroVec impls don't consistently do this, but I'm hoping to move towards always documenting this

**Review by Manishearth — DISMISSED** [review]

**Review by sffc — APPROVED** [review]
LGTM modulo Manish's comment

File: components/uniset/src/ule.rs
Comment (sffc) [reply components/uniset/src/ule.rs:30]: +1 and see https://github.com/unicode-org/icu4x/pull/1121 for an updated safety checklist for ULE

**Review by sffc — APPROVED** [review]

**Review by Manishearth — APPROVED** [review]

**Pull request merged**
